gc/api/include/OMR/GC/Handle.hpp

```diff
@@ -0,0 +1,147 @@
+/*******************************************************************************
+ *  Copyright (c) 2018, 2018 IBM and others
+ *
+ *  This program and the accompanying materials are made available under
+ *  the terms of the Eclipse Public License 2.0 which accompanies this
+ *  distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ *  or the Apache License, Version 2.0 which accompanies this distribution and
+ *  is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ *  This Source Code may also be made available under the following
+ *  Secondary Licenses when the conditions for such availability set
+ *  forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ *  General Public License, version 2 with the GNU Classpath
+ *  Exception [1] and GNU General Public License, version 2 with the
+ *  OpenJDK Assembly Exception [2].
+ *
+ *  [1] https://www.gnu.org/software/classpath/license.html
+ *  [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ *  SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#if !defined(OMR_GC_HANDLE_HPP_)
+#define OMR_GC_HANDLE_HPP_
+
+#include <OMR/TypeTraits.hpp>
+#include <cstddef>
+#include <omrcfg.h>
+
+namespace OMR
+{
+namespace GC
+{
+
+/// A CRTP mixin which adds self() calls which will cast this to t
+template <typename D, typename T>
+class Mixin
+{
+private:
```

Comment:
```suggestion
protected:
```

Reply:
I went with private, so that mixins wouldn't expose the self() method to the base class. I have a suspicion that it would not play out well when a class inherits from multiple CRTP mixins.

Reply:
The thing is, private fields are not accessible by derived classes.